File: stages/stages.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/go-redis/redis"
	"github.com/pkg/errors"
)

/////// Syncer /////////////////

type Syncer interface {
	Set(id, stage string, result *StageResult) error
	Get(id, stage string) (*StageResult, error)
	GetAll(id string) (map[string]*StageResult, error)
	Lock(id, stage string, ttl time.Duration) (bool, error)
	Unlock(id, stage string) error
}
type RedisSyncer struct {
	redis *redis.Client
}

func newRedisSyncer() *RedisSyncer {
	return &RedisSyncer{
		redis: redis.NewClient(&redis.Options{}),
	}
}

func (r *RedisSyncer) Set(id, stage string, res *StageResult) error {
	syncData, err := r.GetAll(id)
	if err != nil {
		return err
	}
	syncData[stage] = res
	jsonVal, err := json.Marshal(syncData)
	if err != nil {
		return errors.Wrap(err, "Unable to marshall stage data")
	}
	r.redis.Set(r.makeKey(id), jsonVal, 0)
	return nil
}

func (r *RedisSyncer) Get(id, stage string) (*StageResult, error) {
	data, err := r.GetAll(id)
	if err != nil {
		return nil, errors.Wrap(err, "Error getting stage data")
	}
	stageData := data[stage]
	return stageData, nil
}

func (r *RedisSyncer) GetAll(id string) (map[string]*StageResult, error) {
	data := make(map[string]*StageResult)
	v, err := r.redis.Get(r.makeKey(id)).Result()
	if err != nil {
		return nil, errors.Wrap(err, "Cannot fetch all data for id:"+id)
	}
	bytes := []byte(v)

	if err := json.Unmarshal(bytes, &data); err != nil {
		return nil, errors.Wrap(err, "Cannot unmarshal stages data")
	}
	return data, nil
}

func (r *RedisSyncer) Lock(id, stage string, ttl time.Duration) (bool, error) {
	key := r.makeKey(id)
	res := r.redis.SetNX(key, true, ttl)
	return true, res.Err()
}

func (r *RedisSyncer) Unlock(id, stage string) error {
	key := r.makeKey(id)
	res := r.redis.Del(key)
	return res.Err()
}

func (r *RedisSyncer) makeKey(s string) string {
	return "stages-" + s
}

/////////////// Syncer end /////////////////////

////////////// Stage Result ///////////////////
type StageResult struct {
	Data   interface{}
	Status string
}

type ResultWriter interface {
	Set(data interface{})
	Bookmark(data interface{})
}

type StageResultWriter struct {
	result   StageResult
	bookmark bool
}

func (st *StageResultWriter) Set(data interface{}) {
	st.result.Data = data
}

func (st *StageResultWriter) Bookmark(data interface{}) {
	st.result.Data = data
	st.bookmark = true
}

////////// Stage result end ////////////////

////////// Single Stage /////////////////
type StageHandler func(rw ResultWriter, data interface{}, seed interface{}) error

type Stage struct {
	name string
	//	result  *StageResult
	handler StageHandler
	ttl     time.Duration
}

type StageConflictError error

// executes stage and returns stageResult, bookmark and error
func (stage *Stage) execStage(syncer Syncer, id string, input, seed interface{}) (*StageResult, bool, error) {
	// Take exclusive lock
	ok, err := syncer.Lock(id, stage.name, stage.ttl)
	if err != nil {
		return nil, false, StageConflictError(err)
	}
	if !ok {
		return nil, false, StageConflictError(fmt.Errorf("cannot get exclusive lock for %s:%s", stage.name, id))
	}
	defer func() {
		if err := syncer.Unlock(id, stage.name); err != nil {
			log.Printf("Cannot unset execution lock %+v", err)
		}
	}()

	// execute stage
	stageResultWriter := &StageResultWriter{}
	if err := stage.handler(stageResultWriter, input, seed); err != nil {
		return nil, false, err
	}
	stageResult := stageResultWriter.result
	stageResult.Status = "done"
	// sync stage data
	if err := syncer.Set(id, stage.name, &stageResult); err != nil {
		return nil, false, errors.Wrap(err, "Cannot sync state result")
	}
	return &stageResult, stageResultWriter.bookmark, nil
}

//////////// Single stage end ///////////

////// Stages ////////////////////

// Stages implements the functionality of idempotent stages
type Stages struct {
	syncer     Syncer
	stages     []*Stage
	lastResult *StageResult
}

// Initialise a Stages struct
func Init(syncer Syncer) *Stages {
	return &Stages{
		syncer: syncer,
		stages: []*Stage{},
	}
}

// Then adds an idempotent step to a Stages object
func (st *Stages) Then(name string, ttl time.Duration, fn StageHandler) *Stages {
	stage := Stage{
		name:    name,
		handler: fn,
		ttl:     ttl,
	}
	st.stages = append(st.stages, &stage)
	return st
}

// Add is an alias for Then
func (st *Stages) Add(name string, ttl time.Duration, fn StageHandler) *Stages {
	return st.Then(name, ttl, fn)
}

// Run executes the correct stage sequence depending on the current saves state
func (st *Stages) Run(id string, seed interface{}) error {
	syncData, err := st.syncer.GetAll(id)
	if err != nil {
		return errors.Wrap(err, "Cannot fetch synced data")
	}
	prevResult := &StageResult{Data: seed}
	for _, stage := range st.stages {
		syncedStageResult, ok := syncData[stage.name]
		if !ok {
			syncedStageResult = &StageResult{}
		}
		if syncedStageResult.Status == "done" {
			prevResult = syncedStageResult
			continue
		}
		stageResult, bookmark, err := stage.execStage(st.syncer, id, prevResult.Data, seed)
		if err != nil {
			return errors.Wrapf(err, "Unable to execute stage %s:%s", stage.name, id)
		}
		prevResult = stageResult
		if bookmark { // Bookmark is an explicit termination to be resumed later
			break
		}
	}
	st.lastResult = prevResult
	return nil
}

func (st *Stages) Result() interface{} {
	return st.lastResult.Data
}

/////////// Stages end //////////

//////// Example /////////
var testSt *Stages
var testOnce sync.Once

func prepareTestStages(syncer Syncer) *Stages {
	return Init(syncer).
		Then("first", 1*time.Second, func(rw ResultWriter, data, seed interface{}) error {
			log.Println("executing first stage")
			str, ok := data.(string)
			if !ok {
				return errors.New("Cannot read input data")
			}
			rw.Set(str + "first")
			return nil
		}).
		Then("second", 1*time.Second, func(rw ResultWriter, data, seed interface{}) error {
			log.Println("executing second stage")
			str, ok := data.(string)
			if !ok {
				return errors.New("Cannot read input data")
			}
			rw.Set(str + "second")
			return nil
		}).
		Then("third", 1*time.Second, func(rw ResultWriter, data, seed interface{}) error {
			log.Println("executing third stage")
			str, ok := data.(string)
			if !ok {
				return errors.New("Cannot read input data")
			}
			rw.Set(str + "third")
			return nil
		})
}

func makeTestStages(syncer Syncer) *Stages {
	testOnce.Do(func() {
		testSt = prepareTestStages(syncer)
	}) // cache
	return testSt // and return
}

func main() {
	st := makeTestStages(newRedisSyncer())
	err := st.Run("random-id1", "seed")
	if err != nil {
		log.Printf("Top level error = %+v\n", err)
		return
	}
	fmt.Printf("result = %+v\n", st.Result())
}

File: stages/stages.rb
# frozen_string_literal: true

require "json"
require "redis"

# class to sync stages to a file
class FileSync
  def set(id, stage, result, status)
    sync_data = get_all(id)
    sync_data[stage] = { "result" => result, "status" => status }
    File.write(id, JSON.generate(sync_data))
  end

  def get(id, stage)
    get_all(id)[stage]
  end

  def get_all(id)
    data = File.read(id)
    JSON.parse(data)
  rescue StandardError
    {}
  end
end

Bookmark = Struct.new(:data)
StageData = Struct.new(:name, :result, :blk, :ttl, keyword_init: true)
StageResult = Struct.new(:data, :status, keyword_init: true)

# class to sync stages to redis
class RedisSync
  def initialize
    @redis = Redis.new
  end

  def set(id, stage)
    sync_data = get_all(id)
    sync_data[stage.name] = stage.result.to_h
    key = make_key(id)
    @redis.set(key, JSON.generate(sync_data))
  end

  def get(id, name)
    get_all(id)[name]
  end

  def get_all(id)
    data = @redis.get(make_key(id))
    JSON.parse(data, symbolize_names: true)
  rescue StandardError
    {}
  end

  def lock(id, stage)
    key = make_lock_key(id, stage.name)
    if @redis.set(key, true, ns: true, ex: stage.ttl)
      return true
    else
      false
    end
  end

  def unlock(id, stage)
    @redis.del(make_lock_key(id, stage.name))
  end

  private

  def make_lock_key(id, name)
    "stages-#{id}-#{name}-lock"
  end

  def make_key(id); "stages-#{id}"; end
end

class StageConflictError < StandardError
  attr_reader :ttl
  def initialize(ttl, stage = nil)
    @ttl = ttl
    @stage = stage
  end

  def to_s
    "stage conflict: #{@stage}: #{@ttl} seconds"
  end
end

# Implements stages execution logic
class Stages
  def initialize(syncer)
    @syncer = syncer
    @stages = []
  end

  def stage(name, time_hint, &blk)
    stage_data = StageData.new(name: name, blk: blk,
       ttl: time_hint, result: StageResult.new(status: "init"))
    @stages.push stage_data
    self
  end
  alias_method :then, :stage

  def run(id, seed = nil)
    sync_data = @syncer.get_all(id) || {}
    merged_stages = @stages.map do |stage|
      sync_stage = sync_data[stage.name.to_sym] || {}
      stage.result = StageResult.new(**sync_stage)
      stage
    end
    run_stages(merged_stages, id, seed)
  end
  
  private

  def run_stages(stages, id, seed = nil)
    stages.reduce(seed) do |input, stage|
      if stage.result.status == "done"
        stage.result.data
      else
        res = exec_stage(id, stage, input, seed)
        if res.is_a? Bookmark
          break res
        else
          next res
        end
      end
    end
  end

  def exec_stage(id, stage, input, seed)
    unless @syncer.lock(id, stage)
      raise StageConflictError.new(stage.ttl / 2)
    end
    blk = stage.blk
    raise "Cannot find handler for stage #{stage.name}" unless blk

    res = blk.call(input, seed)
    sync_res = if res.is_a? Bookmark
                 res.data
                else
                  res
                end
    stage.result = StageResult.new(data: sync_res, status: "done")
    @syncer.set(id, stage)
    res
  ensure
    @syncer.unlock(id, stage)
  end
end

def staged1
  Stages.new(RedisSync.new)
    .then("first", 1) { |data|
      p "doing first"
      # fail "fail at stage 1"
      data + "foo"
    }.then("second", 1) { |data|
      p "doing second"
      # fail "fail at stage 2"
      data + "bar"
    }.then("third", 1) { |data|
      p "doing third"
      # fail "fail at stage 3"
      data + "third"
    }.run("random-id1", "seed")
end

def staged2
  Stages.new(RedisSync.new)
    .then("first", 1) { |data|
      p "doing first"
      Bookmark.new(data + "foo")
    }.then("second", 1) { |_, seed|
      p "doing second"
      Bookmark.new(seed + "bar")
    }.then("third", 1) { |data, _|
      p "doing third"
      data + "third"
    }.run("random-id1", "seed")
end

p staged2
